pkg/adapter/awssqs/adapter.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awssqs
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/knative/pkg/cloudevents"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/signals"
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+const (
+	eventType = "aws.sqs.message"
+)
+
+// Adapter implements the AWS SQS adapter to deliver SQS messages from a
+// an SQS queue to a Sink.
+type Adapter struct {
+
+	// Region is the AWS region
+	Region string
+
+	// QueueUrl is the AWS SQS URL that we're polling messages from
+	QueueUrl string
+
+	// SinkURI is the URI messages will be forwarded on to.
+	SinkURI string
+
+	// CredsFile is the full path of the AWS credentials file
+	CredsFile string
+}
+
+func (a *Adapter) Start(ctx context.Context) error {
+
+	logger := logging.FromContext(ctx)
+
+	stopCh := signals.SetupSignalHandler()
+	cctx, cancel := context.WithCancel(ctx)
+
+	logger.Info("Starting with config: ", zap.Reflect("adapter", a))
+
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigDisable,
+		Config:            aws.Config{Region: aws.String(a.Region)},
+		SharedConfigFiles: []string{a.CredsFile},
+	}))
+
+	q := sqs.New(sess)
+
+	go func() {
```

Comment:
I don't think this function ever exits (even if stopCh is closed or cancel is called).

Reply:
Was the fix pushed?

Reply:
Yes, https://github.com/knative/eventing-sources/pull/124/commits/6ca19c1c24528c28382fcf9f0fd7fe19d7c38583

Reply:
This function (lines 80-109) never exits. The break on line 85 breaks out of the [select statement](https://golang.org/ref/spec#Break_statements) (I think, I'm also pretty new to go), but that just causes the for loop to run again. Because nothing in the function returns and nothing breaks out of the for loop, I think it will run forever.

Reply:
Oh, I did not know break also applied to selects. I'll fix that. Thanks for the catch.